verifier/src/tcb.rs

```diff
@@ -78,13 +131,57 @@ pub struct TcbInfo {
     version: u32,
     issue_date: String,
     next_update: String,
-    fmspc: String,
+    #[serde(with = "hex")]
+    fmspc: [u8; FMSPC_SIZE],
     pce_id: String,
     tcb_type: u32,
     tcb_evaluation_data_number: u32,
     tcb_levels: Vec<TcbLevel>,
 }
 
+impl TcbInfo {
+    /// Get the advisories for the given TCB info
+    ///
+    /// The advisories are the ones corresponding to the TCB levels in `pck_tcb`.
+    ///
+    /// This method maps to steps 3 and 4 from
+    /// <https://api.portal.trustedservices.intel.com/documentation#pcs-tcb-info-v4>
+    ///
+    /// > 3. Go over the sorted collection of TCB Levels retrieved from TCB Info
+    /// >    starting from the first item on the list:
+    /// >   a. Compare all of the SGX TCB Comp SVNs retrieved from the SGX PCK
+    /// >      Certificate (from 01 to 16) with the corresponding values of SVNs
+    /// >      in sgxtcbcomponents array of TCB Level. If all SGX TCB Comp SVNs
+    /// >      in the certificate are greater or equal to the corresponding
+    /// >      values in TCB Level, go to 3.b, otherwise move to the next item
+    /// >      on TCB Levels list.
+    /// >   b. Compare PCESVN value retrieved from the SGX PCK certificate with
+    /// >      the corresponding value in the TCB Level. If it is greater or
+    /// >      equal to the value in TCB Level, read status assigned to this
+    /// >      TCB level. Otherwise, move to the next item on TCB Levels list.
+    /// > 4. If no TCB level matches your SGX PCK Certificate, your TCB Level is
+    /// >    not supported.
+    fn advisories(&self, pck_tcb: &PckTcb) -> Option<Advisories> {
+        // `self` should have been retrieved via
+        // <https://api.trustedservices.intel.com/sgx/certification/v4/tcb?fmspc={}>
+        // and the `pck_tcb.fmspc`. Failure here should rarely happen, but we
+        // still check to ensure the client didn't get mixed up.
+        if self.fmspc != pck_tcb.fmspc {
+            return None;
```

Comment:
i wonder if we should return an actual error? to help distinguish between, your tcb does not match the level of anything in the pck, and there is an fmspc mismatch

Reply:
Updated to return an error